File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums, k) {
        const numToCount = new Map();
        for (const n of nums) {
            const currentCount = numToCount.get(n);
            numToCount.set(n, currentCount === undefined ? 1 : currentCount + 1);
        }

        const buckets = new Array(nums.length + 1).fill(null);
        numToCount.forEach((count, n) => {
            if (buckets[count] === null) {
                buckets[count] = [n];
            } else {
                buckets[count] = [...buckets[count], n];
            }
        });

        const result = new Array(k);
        let resultInd = 0;
        for (let i = buckets.length - 1; i >= 0; i--) {
            if (buckets[i] !== null) {
                for (const n of buckets[i]) {
                    result[resultInd] = n;
                    resultInd++;
                    if (resultInd === k) {
                        return result;
                    }
                }
            }
        }

        return result;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];

        res[0] = nums[0];
        for(int i = 1; i < n - 1; i++) {
            res[i] = res[i - 1] * nums[i];
        }

        int suffix = nums[n - 1];
        res[n - 1] = res[n - 2];
        for(int i = n - 2; i >= 1; i--) {
            res[i] = res[i - 1] * suffix;
            suffix *= nums[i]; 
        }
        res[0] = suffix;

        return res;
    }
}  
